storage/chunked_zstd.go

```diff
@@ -0,0 +1,839 @@
+	}
+
+	buffered := makeBufferedNetworkReader(stream, 64, 16384)
+	defer buffered.Close()
+	mr := multipart.NewReader(buffered, boundary)
+
+	for mc := 0; ; mc++ {
+		p, err := mr.NextPart()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+
+		if mc == len(missingChunks) {
+			return errors.Errorf("too many chunks returned")
+		}
+
+		for _, mf := range missingChunks[mc].Files {
+			if mf.Gap > 0 {
+				limitReader := io.LimitReader(p, mf.Gap)
+				_, err := io.Copy(ioutil.Discard, limitReader)
+				if err != nil {
+					return err
+				}
+				continue
+			}
+
+			limitReader := io.LimitReader(p, mf.Length())
+
+			if err := createFileFromZstdStream(dest, dirfd, limitReader, missingDirsMode, os.FileMode(mf.File.Mode), mf.File, options); err != nil {
+				p.Close()
+				return err
+			}
+		}
+		p.Close()
+	}
+	return nil
+}
+
+func mergeMissingChunks(missingChunks []missingChunk, target int) []missingChunk {
+	if len(missingChunks) <= target {
+		return missingChunks
+	}
+
+	getGap := func(missingChunks []missingChunk, i int) int {
+		prev := missingChunks[i-1].RawChunk.Offset + missingChunks[i-1].RawChunk.Length
+		return int(missingChunks[i].RawChunk.Offset - prev)
+	}
+
+	// this implementation doesn't account for duplicates, so it could merge
+	// more than necessary to reach the specified target.  Since target itself
+	// is a heuristic value, it doesn't matter.
+	var gaps []int
+	for i := 1; i < len(missingChunks); i++ {
+		gaps = append(gaps, getGap(missingChunks, i))
+	}
+	sort.Ints(gaps)
+
+	toShrink := len(missingChunks) - target
+	targetValue := gaps[toShrink-1]
+
+	newMissingChunks := missingChunks[0:1]
+	for i := 1; i < len(missingChunks); i++ {
+		gap := getGap(missingChunks, i)
+		if gap > targetValue {
+			newMissingChunks = append(newMissingChunks, missingChunks[i])
+		} else {
+			prev := &newMissingChunks[len(newMissingChunks)-1]
+			gapFile := missingFile{
+				Gap: int64(gap),
+			}
+			prev.RawChunk.Length += uint64(gap) + missingChunks[i].RawChunk.Length
+			prev.Files = append(append(prev.Files, gapFile), missingChunks[i].Files...)
+		}
+	}
+
+	return newMissingChunks
+}
+
+func retrieveMissingFiles(input *DirectDiffInput, dest string, dirfd int, missingChunks []missingChunk, missingDirsMode os.FileMode, options *archive.TarOptions) error {
+	var chunksToRequest []types.ImageSourceChunk
+	for _, c := range missingChunks {
+		chunksToRequest = append(chunksToRequest, c.RawChunk)
+	}
+
+	// There are some missing files.  Prepare a multirange request for the missing chunks.
+	var stream io.ReadCloser
+	var contentType string
+	var err error
+
+	for stream == nil {
+		stream, contentType, err = input.stream.GetBlobAt(input.ctx, input.srcInfo, chunksToRequest)
+		if err != nil {
+			if _, ok := err.(docker.ErrBadRequest); ok {
+				requested := len(missingChunks)
+				// If the server cannot handle at least 64 chunks in a single request, just give up.
+				if requested < 64 {
+					return err
+				}
```

Comment:
Don't registries that don't support multi-part range request (e.g. gcr.io https://github.com/containerd/stargz-snapshotter/pull/81) always walks the worst code path? (i.e. everytime fails with log n retries)

Reply:
thanks for letting me know about this issue.  I think I could raise the minimum number of chunks to minimize the worst code path in such cases.

Reply:
@giuseppe years later, https://github.com/containers/storage/commit/3bbd19c9dcc0693bd054c51598547b92e52bb811 now loops until `requested == 1`.

Is that something to worry about, or are the registries well-behaved enough that this is fine?

Reply:
@mtrmac I've not seen that behavior with quay.io or docker.io, but hard to say for other registries.

Given that we care more about use cases like composefs now, probably the tradeoff shifted more towards "try harder to use partial images" than it was 4 years ago when it was just an optimization.

@ktock is there any estargz image on `gcr.io` that I can use?

Reply:
(Just to be explicit, I have no opinion on the trade-off and I’m happy to defer to experts; I have just happened to come upon this discussion, and remembered the recent code change.)